src/Shared/JsonSchemaExporter/JsonSchemaExporter.cs

```diff
@@ -0,0 +1,1128 @@
+    /// <exception cref="ArgumentNullException">One of the specified parameters is <see langword="null" />.</exception>
+    /// <exception cref="NotSupportedException">The <paramref name="typeInfo"/> parameter contains unsupported exporterOptions.</exception>
+    [RequiresUnreferencedCode(RequiresUnreferencedCodeMessage)]
+    public static JsonNode GetJsonSchemaAsNode(this JsonTypeInfo typeInfo, JsonSchemaExporterOptions? exporterOptions = null)
+    {
+        _ = Throw.IfNull(typeInfo);
+        ValidateOptions(typeInfo.Options);
+
+        exporterOptions ??= JsonSchemaExporterOptions.Default;
+        return MapRootTypeJsonSchema(typeInfo, exporterOptions);
+    }
+
+    [RequiresUnreferencedCode(RequiresUnreferencedCodeMessage)]
+    private static JsonNode MapRootTypeJsonSchema(JsonTypeInfo typeInfo, JsonSchemaExporterOptions exporterOptions)
+    {
+        GenerationState state = new(exporterOptions, typeInfo.Options);
+        JsonSchema schema = MapJsonSchemaCore(ref state, typeInfo);
+        return schema.ToJsonNode(exporterOptions);
+    }
+
+    [RequiresUnreferencedCode(RequiresUnreferencedCodeMessage)]
+    private static JsonSchema MapJsonSchemaCore(
+        ref GenerationState state,
+        JsonTypeInfo typeInfo,
+        Type? parentType = null,
+        JsonPropertyInfo? propertyInfo = null,
+        ICustomAttributeProvider? propertyAttributeProvider = null,
+        ParameterInfo? parameterInfo = null,
+        bool isNonNullableType = false,
+        JsonConverter? customConverter = null,
+        JsonNumberHandling? customNumberHandling = null,
+        JsonTypeInfo? parentPolymorphicTypeInfo = null,
+        bool parentPolymorphicTypeContainsTypesWithoutDiscriminator = false,
+        bool parentPolymorphicTypeIsNonNullable = false,
+        KeyValuePair<string, JsonSchema>? typeDiscriminator = null,
+        bool cacheResult = true)
+    {
+        Debug.Assert(typeInfo.IsReadOnly, "The specified contract must have been made read-only.");
+
+        JsonSchemaExporterContext exporterContext = state.CreateContext(typeInfo, parentPolymorphicTypeInfo, parentType, propertyInfo, parameterInfo, propertyAttributeProvider);
+
+        if (cacheResult && typeInfo.Kind is not JsonTypeInfoKind.None &&
+            state.TryGetExistingJsonPointer(exporterContext, out string? existingJsonPointer))
+        {
+            // The schema context has already been generated in the schema document, return a reference to it.
+            return CompleteSchema(ref state, new JsonSchema { Ref = existingJsonPointer });
+        }
+
+        JsonSchema schema;
+        JsonConverter effectiveConverter = customConverter ?? typeInfo.Converter;
+        JsonNumberHandling effectiveNumberHandling = customNumberHandling ?? typeInfo.NumberHandling ?? typeInfo.Options.NumberHandling;
+
+        if (!IsBuiltInConverter(effectiveConverter))
+        {
+            // Return a `true` schema for types with user-defined converters.
```

Comment:
Can you help me better understand this part? Adding a custom converter to something means that it no longer participates in schema generation?

Reply:
A custom converter means that we are not able to reason about what schema the type may support, so we use the schema allowing any JSON value.